### PR TITLE
Bacpop-98 Persist project data before analysis runs 

### PR DESCRIPTION
**Pull request by absternator**
This ticket allows users to persist their project data before analysis is run.  
The following is done 
- endpoint to save sketch data 
- Edit getProject endpoint! This is the main work here and allows to get project info at all stages
- redis store updates
- unit tests

Note: The backend is the main focus of this ticket... the frontend has just been updated to ensure to it all works (because of rewrite). Thus i wouldnt be worried about reviewing FE code... ensuring it workings should be enough

**Dev Testing **
Can be tested via postman or just the beebop client....Test the following:
- Ensure you can save sketch data
- Be able to query an empty project 
- Be able to query project with files before the run has happened
- Be able to get data correctly after project has been run

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
[the remark this answers is not in the document]
Comment (absternator) [reply .vscode/launch.json:1]: removed as requested

**Review by absternator — COMMENTED** [review]
All FE related things are great and will be taken into account with the NEW FE

File: app/server/src/controllers/projectController.ts [hunk 2]
```diff
@@ -1,80 +1,107 @@
...
-                }
-            });
-        },
+    async getProject(request, response, next) {
+      await asyncHandler(next, async () => {
+        const { projectId } = request.params;
+        const { redis } = request.app.locals;
+        const store = userStore(redis);
+        const baseProjectInfo = await store.getBaseProjectInfo(projectId);
+        const projectSplitSampleIds = await store.getProjectSplitSampleIds(projectId);
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/server/src/controllers/projectController.ts:41]: pretty much haha

File: app/server/src/db/userStore.ts [hunk 3]
```diff
@@ -48,6 +49,10 @@ export class UserStore {
     async getProjectHash(request, projectId: string) {
          return await this._redis.hget(this._projectKey(projectId), "hash")
     }
+    
+    async getBaseProjectInfo(projectId: string) {
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/server/src/db/userStore.ts:53]: done... bit weird though had to convert to unkown first as the returned type from `hgetAll` is `Record<string, string>`

File: app/server/src/types/models.ts [hunk 4]
```diff
@@ -7,4 +7,17 @@ export interface AMR {
...
+    filename: string;
+}
+
+export interface ProjectSample {
+    hash: string;
+    filename: string;
+    amr: AMR;
+    sketch: Record<string, unknown>;
+    cluster?: number;
+  }
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/server/src/types/models.ts:23]: have applied suggestion... but i disagree its an implementation detail as now we know what exactly we are putting in redis and whats coming out... so dont have to scan code and and be unsure what redis will be returning

File: app/server/src/utils/jsonUtils.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+    public static safeStringify(obj): string | null {
+        try {
+            return JSON.stringify(obj);
+        } catch (error) {
+            console.error("Error stringifying object:", error);
+            return null;
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/server/src/utils/jsonUtils.ts:20]: sometimes i think when pinging api and the data didnt exist.. i have made it so it just throws now of fail

File: app/server/src/controllers/projectController.ts [hunk 6]
```diff
@@ -1,80 +1,107 @@
...
+          projectId,
+          projectSplitSampleIds,
+          apiData
+        );
+        return sendSuccess(response, {
+          ...baseProjectInfo,
+          ...apiData,
+          samples: responseSamples,
+        });
+      });
```
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply app/server/src/controllers/projectController.ts:78]: Not for me!

**Review by EmmaLRussell — APPROVED** [review]
Still not mad keen on `SplitSampleId`, but I'll get used to it 😆

**Pull request merged**